### ISSUE TITLE
Support efficient operations for large matrix 

### ISSUE THREAD

**Issue by erizmr** [taichi-dev/taichi#2696]
**Concisely describe the proposed feature**
Support efficient operations (e.g. matmul ) for large matrix. The current `ti.Matrix` is designed for small matrices, therefore the compilation is extreme slow for even a moderate large matrix.

p.s. Also an error shown in following line is raised for large matrix when materializing a large matrix:
```
RuntimeError: [struct_llvm.cpp:run@307] Assertion failure: (int)snodes.size() <= taichi_max_num_snodes
```
To reproduce:
```
import taichi as ti

ti.init()

a = ti.Matrix.field(36, 18, dtype=ti.f32, shape=())
b = ti.Matrix.field(18, 36, dtype=ti.f32, shape=())

@ti.kernel
def test():
    # print(b[None] @ a[None])
    print(b[None])

test()
```

**Describe the solution you'd like (if any)**
Store the elements of large matrices (`n*m>32`) in a `ti.field` with large `shape=(n, m)` but not matrices size. Extend the current functions of `ti.Matrix` for large matrices based on the field. 

**Additional comments**
Any discussion and suggestions are welcome!
Personally I think although the users who need a large matrix can store it in a field with large shape directly, it requires the users to implement basic matrix operations such as `matmul` by themselves.  It is not a huge work for professional users but it may potentially degrade new user's experience when they just want to seek a quick off-the-shelf solution.

Related issues and discussions #983 #833

**Comment by eatcosmos** [comment]
I would like to help!

**Comment by lucywsq** [comment]
Hi, @eatcosmos if you’re still working on this issue, please let us know, thanks!

**Comment by eatcosmos** [comment]
> Hi, @eatcosmos if you’re still working on this issue, please let us know, thanks!

I'm sorry long time no reply. i quit this，maybe need more background knowledge，other student can help

**Comment by Hilbert-Yaa** [comment]
Hi, if @eatcosmos decides to quit, I am interested to set my hands to this issue, if it's eligible could you please assign that to me? : ) I need some directions. 
I read through the Python code for ti.Matrix in Taichi(./python/taichi/lang/Matrix.py), and as far as I know(pls correct me if I got something wrong) it seems that it's pure Python implementation? (e.g. operations such as ti.Matrix.__matmul__ are implemented in nested loops of Python -> and unrolling will be applied-> slow compilation & execution).
Do you think it's feasible to rewrite an alternative version where ti.Matrix is just a Python wrapper of its cpp impl. ? (I believe there's plentiful existing work e.g. in linalg.h already, and some interfaces have been done.)

**Comment by archibate** [comment]
>  an alternative version where ti.Matrix is just a Python wrapper of its cpp impl

Exactly, I saw recently `ti.Struct` has been added`, why not add `ti.Array` for dynamic-indexable array too? Their cpp equivalent are `std::tuple` and `std::array`, where `std::array` allows dynamic indexing using `a[i]` and can be large but all elements has to be same type, and `std::tuple` only supports compile-time index like `std::get<i>(a)` but allows each element to be different type.

**Comment by Hilbert-Yaa** [comment]
Thanks, I'll start working on it right this weekend. 😆

**Comment by strongoier** [comment]
Hi @Hilbert-Yaa. Welcome to the Taichi community! In fact, `ti.Vector` is on its way to become a dynamic-indexable array, and we already have `TensorType` in Taichi core now. Also, putting all `Matrix`-related implementations into Taichi core is on our current roadmap. We're glad that you would like to help. Stay in touch and we can discuss more implementation details when you're available.

**Comment by Hilbert-Yaa** [comment]
> Hi @Hilbert-Yaa. Welcome to the Taichi community! In fact, `ti.Vector` is on its way to become a dynamic-indexable array, and we already have `TensorType` in Taichi core now. Also, putting all `Matrix`-related implementations into Taichi core is on our current roadmap. We're glad that you would like to help. Stay in touch and we can discuss more implementation details when you're available.

Thanks for your warm welcome @strongoier ! As planned I'll begin to figure things this weekend. Stay in touch 🌟.
